storage/paths/db_index.go

```diff
@@ -0,0 +1,882 @@
+			}
+
+			result = append(result, storiface.SectorStorageInfo{
+				ID:         storiface.ID(row.StorageId),
+				URLs:       urls,
+				BaseURLs:   burls,
+				Weight:     row.Weight * 0,
+				CanSeal:    row.CanSeal,
+				CanStore:   row.CanStore,
+				Primary:    false,
+				AllowTypes: splitString(row.AllowTypes),
+				DenyTypes:  splitString(row.DenyTypes),
+			})
+		}
+	}
+
+	return result, nil
+}
+
+func (dbi *DBIndex) StorageInfo(ctx context.Context, id storiface.ID) (storiface.StorageInfo, error) {
+
+	var qResults []struct {
+		Urls       string
+		Weight     uint64
+		MaxStorage uint64
+		CanSeal    bool
+		CanStore   bool
+		Groups     string
+		AllowTo    string
+		AllowTypes string
+		DenyTypes  string
+	}
+
+	err := dbi.harmonyDB.Select(ctx, &qResults,
+		"SELECT urls, weight, max_storage, can_seal, can_store, groups, allow_to, allow_types, deny_types "+
+			"FROM StorageLocation WHERE storage_id=$1", string(id))
+	if err != nil {
+		return storiface.StorageInfo{}, xerrors.Errorf("StorageInfo query fails: %v", err)
+	}
+
+	var sinfo storiface.StorageInfo
+	sinfo.ID = id
+	sinfo.URLs = splitString(qResults[0].Urls)
+	sinfo.Weight = qResults[0].Weight
+	sinfo.MaxStorage = qResults[0].MaxStorage
+	sinfo.CanSeal = qResults[0].CanSeal
+	sinfo.CanStore = qResults[0].CanStore
+	sinfo.Groups = splitString(qResults[0].Groups)
+	sinfo.AllowTo = splitString(qResults[0].AllowTo)
+	sinfo.AllowTypes = splitString(qResults[0].AllowTypes)
+	sinfo.DenyTypes = splitString(qResults[0].DenyTypes)
+
+	return sinfo, nil
+}
+
+func (dbi *DBIndex) StorageBestAlloc(ctx context.Context, allocate storiface.SectorFileType, ssize abi.SectorSize, pathType storiface.PathType) ([]storiface.StorageInfo, error) {
+	var err error
+	var spaceReq uint64
+	switch pathType {
+	case storiface.PathSealing:
+		spaceReq, err = allocate.SealSpaceUse(ssize)
+	case storiface.PathStorage:
+		spaceReq, err = allocate.StoreSpaceUse(ssize)
+	default:
+		return nil, xerrors.Errorf("Unexpected path type")
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("estimating required space: %w", err)
+	}
+
+	var checkString string
+	if pathType == storiface.PathSealing {
+		checkString = "can_seal = TRUE"
+	} else if pathType == storiface.PathStorage {
+		checkString = "can_store = TRUE"
+	}
+
+	var rows []struct {
+		StorageId  string
+		Urls       string
+		Weight     uint64
+		MaxStorage uint64
+		CanSeal    bool
+		CanStore   bool
+		Groups     string
+		AllowTo    string
+		AllowTypes string
+		DenyTypes  string
+	}
+
+	sql := fmt.Sprintf(`SELECT storage_id, 
+								urls, 
+								weight,
+								max_storage, 
+								can_seal, 
+								can_store, 
+								groups, 
+								allow_to, 
+								allow_types, 
+								deny_types 
+						 FROM storagelocation 
+						 WHERE %s and available >= $1
+						 and NOW()-last_heartbeat < $2 
+						 and heartbeat_err is null
+						order by (available::numeric * weight) desc`, checkString)
+	err = dbi.harmonyDB.Select(ctx, &rows,
```

Comment:
no need to build a query. Just pass in (pathType==storiface.PathSealing) as $4
and ($4 AND can_seal OR NOT $4 AND can_store)

Reply:
Whats the downside of building a query?

> Just pass in (pathType==storiface.PathSealing) as $4
and ($4 AND can_seal OR NOT $4 AND can_store)

This assumes that either CanSeal or CanStore will be set when actually both of them could be set

Reply:
Query building leads to SQL injections if done incorrectly (this one looks fine).
Honestly I thought I had put things in place to block query building (harmonydb.rawString). 

I attempted to replicate the above logic with: NOT $4
       The point is to say that the logic could live in the SQL.

Reply:
done